snapcraft/internal/pluginhandler/stage_package_grammar/_on.py

```diff
@@ -0,0 +1,128 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+
+from . import (
+    process_grammar,
+    StagePackageSyntaxError
+)
+
+
+class OnStatement:
+    """Process an 'on' statement in the stage packages grammar.
+
+    Example:
+    >>> import tempfile
+    >>> from snapcraft import repo, ProjectOptions
+    >>> with tempfile.TemporaryDirectory() as cache_dir:
+    ...     ubuntu = repo.Ubuntu(cache_dir)
+    ...     options = ProjectOptions(target_deb_arch='i386')
+    ...     clause = OnStatement('on amd64', ['foo'], options, ubuntu)
+    ...     clause.add_else(['bar'])
+    ...     clause.process()
+    {'bar'}
+    """
+
+    def __init__(self, on, body, project_options, ubuntu):
```

Comment:
Alright, I'm making this change now. By the way, yes, Item 21 does indeed recommend using keyword-only arguments to enforce clarity, but note that he doesn't necessarily recommend doing it for all arguments, just ones that aren't clear (e.g. his example only made two out of four arguments keyword-only as they were both booleans and "it's easy to confuse the position of the two Boolean arguments").